[3 - 4단계 방탈출 예약 대기] 벨로(방예혁) 미션 제출합니다.

YehyeokBang opened the pull request:
안녕하세요 엘리!

이번에도 잘 부탁드립니다. 😄 

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
  - <img width="209" alt="image" src="https://github.com/user-attachments/assets/44a06de5-0c6d-48a2-8009-2d149072e0dd" />
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [벨로의 프롤로그](https://prolog.techcourse.co.kr/studylogs/4381)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

- 하나의 일만 하도록 하려고 노력했습니다.
- Waiting과 Member가 협력하여 승인 및 취소 권한을 확인하는 것처럼 꺼내서 작업하는 것을 지양했습니다.

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

YehyeokBang committed:
docs: 구현할 기능 목록 작성

- 3단계 요구사항에 맞게 구현 기능 목록을 재작성합니다.

YehyeokBang committed:
feat: 예약 화면에 대기 버튼 추가

- 이미 예약된 시간대인 경우 대기 버튼이 활성화됩니다.

YehyeokBang committed:
feat: 예약 대기 기능 구현

- 이미 예약이 존재하는 시간대에 대기를 요청할 수 있습니다.

YehyeokBang committed:
feat: 대기 순번 조회 기능 구현

- 내 예약 목록과 함께 대기 현황(순번 포함)을 반환합니다.

YehyeokBang committed:
feat: 예약 대기 취소 기능 구현

YehyeokBang committed:
refactor: 저장소 클래스 패키지 분리

- 별도의 저장소 패키지를 추가하고 이동합니다.

YehyeokBang committed:
refactor: 사용자 이메일 값 객체 class로 변환

YehyeokBang committed:
refactor: 클래스 레벨 final 키워드 제거

YehyeokBang committed:
refactor: 권한 클래스명 수정

- 더욱 명확하게 변경합니다.

YehyeokBang committed:
fix: 관리자 권한 확인 오류 수정

- 논리 오류를 수정합니다.

YehyeokBang committed:
refactor: 서비스 객체 분리

- Command와 Query 형태로 분리합니다.

* refactor: 예약 시간 관련 서비스 객체 분리

- Command와 Query로 분리합니다.

* refactor: 테마 관련 서비스 객체 분리

- Command와 Query로 분리합니다.

* refactor: 예약 관련 서비스 객체 분리

- Command와 Query로 분리합니다.

* refactor: DTO 패키지 이동

* refactor: 예약 대기 관련 서비스 객체 분리

- Command와 Query로 분리합니다.
- 테스트 코드를 추가합니다.

* refactor: 사용자 관련 서비스 객체 분리

- Command와 Query로 분리합니다.

YehyeokBang committed:
refactor: 상수 분리 및 메서드 순서 정렬

YehyeokBang committed:
fix: 테스트 메서드 내 예외 검증 메시지 수정

YehyeokBang committed:
refactor: 내 예약 목록 기능 리팩터링

- 가독성을 위한 메서드를 분리합니다.
- 클라이언트의 쉬운 분리를 위해 타입 필드를 추가합니다.

YehyeokBang committed:
refactor: 예약 생성 기능 리팩터링

- 가독성을 위한 메서드를 분리합니다.

YehyeokBang committed:
docs: 구현할 기능 목록 작성

- 4단계 구현 기능 목록을 작성합니다.

YehyeokBang committed:
feat: 예약 대기 관리 화면 추가

- 관련 js 파일도 추가합니다.

YehyeokBang committed:
feat: 예약 대기 목록 조회 기능 구현

- 관리자가 대기 목록을 조회할 수 있습니다.

YehyeokBang committed:
feat: 예약 대기 취소(거절) 기능 구현

- 관리자가 대기 목록 중 원하는 항목을 거절할 수 있습니다.

YehyeokBang committed:
feat: 예약 대기 승인 기능 구현

- 관리자가 대기 목록 중 선택한 항목을 승인합니다.
- 이전 존재하던 예약이 취소되었고, 첫 번째 대기인 경우에만 승인 가능합니다.

YehyeokBang committed:
feat: 중복 대기 검증 기능 구현

- 같은 정보로 예약 또는 대기가 있는 경우 대기를 신청할 수 없습니다.

YehyeokBang committed:
feat: 도메인 예외 핸들링 구현

YehyeokBang committed:
refactor: 클라이언트 코드 완료된 TODO 제거

YehyeokBang committed:
refactor: 더미 데이터 및 css 수정

- 더미 데이터를 한글로 변경합니다.
- 인기 테마 화면의 이미지 크기를 조정합니다.

YehyeokBang committed:
feat: 예약 대기 승인 및 취소 기능 에러 alert 구현

YehyeokBang committed:
refactor: 검색 기능 리팩터링

- 미선택 select option을 추가합니다.
- 모든 선택 옵션이 선택으로 변경됩니다.

YehyeokBang committed:
refactor: 대기 순서 확인 방식 변경

- id가 아닌 createAt을 기준으로 순서를 결정합니다.
- createAt, updatedAt을 가진 기반 엔티티를 추가합니다.

YehyeokBang committed:
fix: 순위 정보가 매번 달라지던 오류 수정

- 같은 순위일 경우를 위해 보조 기준을 추가합니다.
- limit 대신 Pageable 객체를 사용합니다.

YehyeokBang committed:
feat: 예약 시간 출력 포맷 추가

- 불필요했던 초 단위를 제거하고 반환합니다.

YehyeokBang committed:
test: 테스트 전 id 초기화 전략 테스트 분리

- 기존 id 확인을 하드코딩으로 하던 방식에서 영속화된 객체에게 가져와서 확인하도록 변경합니다.
- 초기화 검증은 추상 클래스를 테스트 하도록 분리합니다.

YehyeokBang committed:
fix: 더미 데이터에 있던 중복 예약 제거

YehyeokBang committed:
refactor: 에러 메시지 alert 구현

YehyeokBang committed:
feat: 과거 일시 대기 검증 구현

- 지난 날에 대기를 요청할 수 없습니다.

YehyeokBang committed:
feat: 관리자 화면에 대기 화면 이동 버튼 추가

- 기타 어색하던 버튼의 경로를 수정합니다.

YehyeokBang committed:
refactor: 가독성 개선을 위한 메서드 분리

- 불필요한 쿼리를 제거합니다.

YehyeokBang committed:
refactor: 서비스 코드 리팩터링

- 비즈니스 로직을 더욱 이해하기 쉽게 개선합니다.
- 관련 테스트 코드도 수정합니다.

YehyeokBang committed:
fix: 자신의 대기를 취소하지 못하던 오류 수정

YehyeokBang committed:
fix: 반환 시간 타입 수정

- 불필요한 초 단위를 제거합니다.

YehyeokBang committed:
docs: 구현 기능 목록 작성

- 추가된 기능을 정리하여 작성합니다.

YehyeokBang committed:
test: 예약 대기 체험을 위한 더미 데이터 추가

- 예약 대기 데이터를 추가합니다.

YehyeokBang committed:
test: 예약 대기 관련 테스트 작성

YehyeokBang committed:
refactor: 가독성 개선을 위한 리팩터링

YehyeokBang committed:
fix: 사용자가 자신의 대기를 승인할 수 있던 오류 수정

- 승인 권한과 취소 권한 확인 메서드를 분리합니다.

YehyeokBang committed:
test: 사용자 객체 테스트 작성

YehyeokBang committed:
test: 테마 객체 테스트 작성

YehyeokBang commented:
@coderabbitai full review

YehyeokBang commented on README.md:
```diff
@@ -10,17 +10,53 @@
 
 [Quick Start](#quick-start) 섹션으로 이동하여 애플리케이션을 실행하고 기능을 체험할 수 있습니다.
 
+# 추가된 화면 및 기능
```
## 이전 리뷰

> 또는 서비스 클래스 자체를 조회성 (query) / 명령성 (command) 으로 분리해서 오버라이드할 필요 없게끔 구현하는 것도 좋아합니다.
>
> [트랜잭션 클래스 레벨 vs 메서드 레벨 리뷰](https://github.com/woowacourse/spring-roomescape-waiting/pull/195#discussion_r2094145009)

엘리의 `조회(Query)와 명령(Command) 책임 분리` 의견을 듣고,
서비스 계층을 조회성과 명령성으로 명확히 나누어 구현해보았습니다.

특히 명령성 서비스는 주로 객체 생성이나 상태 변경(대기 -> 예약 승격 등)과 같은 책임을 갖도록 작게 분리했고,
조회성 서비스는 도메인 단위로 나누어 다루도록 구성했습니다.

**느낀 장점**
- 명령성 서비스는 여러 도메인의 의존성을 가질 수 있지만, 기능별로 하나의 책임만 수행하여 이해하기 쉽다.
  - 테스트 코드도 작성하기 훨씬 수월했습니다.
- 클래스명만 보아도 어떤 도메인에서 어떤 작업을 수행하는지 쉽게 파악할 수 있어 가독성이 좋아진 것 같다.
- 조회성 서비스는 필요한 의존성이 적고, 읽기 전용 기능만 포함하므로 코드가 단순해졌다.

**느낀 단점**
- 클래스 수가 많아진다.
  - 다만, 더욱 또렷해졌다고 생각하면 장점으로 볼 수도 있다. 기존 XXXService는 모든 로직을 다 담당하고 있어서 수정하기 어려웠다.

재밌고 좋은 시도라고 느꼈습니다.
이번 Command, Query 분리에 관한 엘리의 피드백도 받고 싶습니다! 😄

yebink replied:
명령성 서비스도 생성인지 삭제인지에 따라 세세하게 잘 분리하셨네요!
서비스의 크기도 크지 않고, 한 가지 기능에만 집중할 수 있어서 좋네요 😄 

`CreateXXService` / `DeleteXXService` -> `XXCommandService` 로 합쳐서 관리하면 
클래스가 많아진다는 단점은 좀 상쇄되는 것 같습니다 ㅎㅎ

하지만 지금 구조도 충분히 좋네요 👏

YehyeokBang replied:
맞아요! 한 가지 기능에 집중할 수 있는 것이 또렷하게(명확하게) 볼 수 있어서 좋은 것 같아요.

> CreateXXService / DeleteXXService -> XXCommandService 로 합쳐서 관리하면
클래스가 많아진다는 단점은 좀 상쇄되는 것 같습니다 ㅎㅎ

혹시 이 부분은 어떤 구조를 말씀하시는 걸까요?

```java
// 1
class XXCommandService {
    
    private final CreateXXService;
    private final DeleteXXService;
}

// 2
class XXCommandService {
    
    // Create, Delete 합치기
}
```

1번은 Facade처럼 느껴지고, 어떤 명령이 있는지 모아서 볼 수 있어서 좋을 것 같네요. 결국 서비스를 호출하게 되는 Controller 입장에서는 의존 복잡도를 줄일 수 있겠네요.

2번은 어떤 명령이 있는지는 메서드 레벨에서 확인이 필요하지만, 클래스 수를 획기적으로 줄일 수 있을 것 같네요. 규모가 복잡하지 않다면, 실용적일 것 같아요.

yebink replied:
2번이었습니다 😄

YehyeokBang commented on src/test/java/roomescape/application/AbstractServiceIntegrationTestTest.java:
```diff
@@ -0,0 +1,44 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.JdbcTemplate;
+
+class AbstractServiceIntegrationTestTest extends AbstractServiceIntegrationTest {
```
## 이전 리뷰

> 초기화 로직을 검증하기 위한 별도 테스트를 구성하는 것이 협업의 측면에선 더 명확한 코드일 것 같네요.
>
> [테스트 단언문에서 id 하드코딩 리뷰](https://github.com/woowacourse/spring-roomescape-waiting/pull/195#discussion_r2094144963)

엘리의 의견에 동의합니다! 다시 생각해보니 테스트마다 초기화가 잘 되었는지 확인한다는 말 같기도 하네요.
모두 수정한 후 초기화 로직을 별도로 검증할 수 있게 `AbstractServiceIntegrationTestTest`를 추가했습니다. 😄

yebink replied:
💯

YehyeokBang commented on src/main/java/roomescape/application/reservation/command/WaitingPromotionService.java:
```diff
@@ -0,0 +1,98 @@
+        this.clock = clock;
+    }
+
+    public Long approve(Long waitingId, Long memberId) {
+        Member member = getMember(memberId);
+        Waiting waiting = getWaiting(waitingId);
+        validateControlPermission(member, waiting);
+        validateReservationIsAvailable(waiting);
+        validateWaitingIsFirst(waiting);
+        Reservation savedReservation = promoteWaitingToReservation(waiting);
+        return savedReservation.getId();
+    }
```
자동 승인을 구상하다가 고민(우려?)이 생겨서 우선 수동 승인으로 기능을 구현했습니다.

### 자동 승인?

> 예약이 취소되었을 때, 그 예약을 기다리던 대기 중 가장 빠른 요청을 자동으로 예약으로 전환하는 기능

이 기능을 구현하려면 결국 예약 취소 로직 안에서 대기 목록을 확인하고,예약 생성까지 처리하는 로직이 추가되어야 합니다.
하지만 이 방식은 자연스럽게 예약 도메인이 대기 도메인을 직접 알아야만 합니다. (강한 결합..?)
처음엔 예약과 대기는 최대한 분리하면서 구성하려 했는데, 알아야만 하는 건가? 라는 생각도 들었습니다.
(예약과 대기는 다르다. 예약은 대기가 추가되어도 알아서 잘 살고, 대기가 그냥 기다리는..? 😢)

### 더 복잡해지면?

예를 들어, 결제 로직이 추가된다면,
`예약 취소 시 결제 취소 → 대기 확인 및 자동 예약 → 예약 생성 → 결제 생성`
처럼 점점 다양한 도메인이 연결될 것 같습니다.

이때마다 각 도메인끼리 직접 호출하게 되면 서로 모르면 좋을 객체들이 알 수 밖에 없는 구조?
중간에 조율해주는 객체가 필요한가? 아니면 원래 이런가..? 라는 고민이 생겼습니다.

엘리의 경험과 의견이 궁금합니다!!

yebink replied:
오 ㅎㅎ 제 경험으로는 `원래 이런가..?` 가 맞는 것 같습니다 ㅎㅎㅎ

도메인 복잡도를 낮추기 위해 수동 승인으로 구현하신 점은 아주 좋은 선택이네요 👍 

현업에서도 서비스는 한 번에 완성되지 않습니다. 기본적인 기능을 먼저 오픈하고 점점 고도화를 시켜 나가게 되죠!
여유가 되신다면 `자동 승인으로 변경`하는 경험을 해보는 것도 좋을 것 같네요! (필수는 아닙니다~)

YehyeokBang replied:
감사합니다! 😄 

도메인 복잡도를 낮추기 위해 `수동 승인`을 선택했지만,
제가 관리자라면, `수동 승인` 보다는 `자동 승인`을 더 선호할 것 같다고 생각했어요.

그래서 도메인 간 의존도는 줄이면서 `자동 승인`을 구현하기 위해 `이벤트 발행-구독`을 활용했습니다.

[feat: 예약 취소 시 대기 자동 승인 기능 구현](https://github.com/woowacourse/spring-roomescape-waiting/pull/261/commits/a44ecfe9180a8f896ca026708b673564df8d687f)

yebink replied:
와우우 👏 👏 👏 👏 👏

YehyeokBang commented on src/main/java/roomescape/domain/reservation/Waiting.java:
```diff
@@ -0,0 +1,90 @@
+        }
+    }
+
+    public boolean hasApproveControlPermission(Member member) {
+        return member.isAdmin();
+    }
+
+    public boolean hasCancelControlPermission(Member member) {
+        return isOwner(member) || member.isAdmin();
+    }
```
대기 승인이나 취소를 구현할 때
`waiting.getMember().getId().equals(member.getId())`처럼 ID를 꺼내 비교하기보다,
연관된 `Member` 객체를 직접 받아 비교하는 방식이 더 객체지향적이라고 판단했습니다.

그래서 `Waiting` 도메인에 `hasApproveControlPermission`, `hasCancelControlPermission` 같은 메서드를 추가하여,
권한 판단 책임을 객체 내부로 위임하는 구조를 선택했습니다.

### 궁금한 점

이처럼 비교 대상 Member 객체(연관된 객체 또는 다른 엔티티 등)를 외부에서 주입받아 비교할 경우,
의도치 않은 사이드 이펙트나 결합이 발생할 가능성은 없을지 궁금합니다.
특히 도메인 객체 간 비교에서 신중해야 할 부분이 있다면 조언 부탁드립니다! 😄

yebink replied:
해당 방식이 더 객체지향적이라는 부분에 동의합니다.

하지만 `waiting.hasApproveControlPermission` 이런 식으로 호출이 되는데,
permission을 가진 주체는 waiting이 아니라 member 여서 살짝 어색한 것 같네요!
관련한 부분을 고민해보시면 좋을 것 같습니다 ㅎㅎ
의도치 않은 사이드 이펙트 / 결합 은 크게 안 보이네요! 👏

YehyeokBang replied:
좋은 의견 감사합니다 👍 👍 

어색함은 메서드를 읽을 때, `권한의 주체`와 `객체의 위치`가 바뀐 것처럼 보이기 때문에 생기는 것 같아요.

## 권한 주체가 판단하도록 변경

`member.canApprove(waiting)`처럼 주체인 `Member`에게 권한 여부를 묻게 하는 방식도 고민해봤습니다.

읽는 방향이 자연스럽고, 권한 판단의 책임이 주체에 있음이 드러나요.
그러나 `Member`가 `Waiting`, `Reservation` 등을 알아야 하고,
`Waiting 승인 및 대기`라는 비즈니스 규칙이 Waiting이 아닌 외부로 흩어지는 느낌을 받았어요.

## 권한 정책 클래스로 분리

`ApprovalPermission.canApprove(member, waiting)`처럼 분리하는 방법도 고민했습니다.

책임 분리가 명확해지고 테스트 용이성이 높아질 것 같아요.
그러나 마찬가지로 비즈니스 로직이 객체 바깥으로 이동해 객체지향적(?)인 응집력은 다소 줄어드는 것 같아요.

## 유지하고 메서드명 개선

기존 `waiting.hasApproveControlPermission(member)`의 방식은 규칙이 `Waiting` 내부에 존재하여 응집도를 높입니다.

다만 문장 구조상 약간 어색함이 있었는데, 
`canBeApprovedBy(member)`처럼 메서드명을 바꿔 표현의 어색함을 푸는 것은 어떨까요??

- `Waiting`은 이미 `Member`를 알고 있는 도메인 객체이다.
- `Member`가 다른 도메인을 알아야 할 이유를 줄일 수 있다.

[refactor: 대기 제어 확인 메서드명 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/261/commits/261df390ff651d7ec45824449910bd4b93b18e4f)

yebink replied:
훨씬 읽기 편리하네요! 💯

YehyeokBang reviewed (commented):
항상 꼼꼼한 리뷰 감사해요 엘리! 😄 
이번에도 잘 부탁드립니다. 👍

YehyeokBang committed:
fix: 회원가입 요청 DTO 검증 추가

YehyeokBang commented on src/main/java/roomescape/application/member/command/dto/RegisterCommand.java:
```diff
@@ -0,0 +1,8 @@
+        String email,
+        String password,
+        String name
+) {
```
수정할게

YehyeokBang commented on src/main/java/roomescape/application/reservation/command/WaitingPromotionService.java:
```diff
@@ -0,0 +1,98 @@
+        Reservation reservation = createReservationBy(waiting);
+        reservation.validateReservable(LocalDateTime.now(clock));
+        return reservationRepository.save(reservation);
+    }
```
학습해볼게

YehyeokBang commented on src/main/java/roomescape/application/reservation/command/DeleteReservationService.java:
```diff
@@ -0,0 +1,20 @@
+    public void removeById(Long reservationId) {
+        reservationRepository.deleteById(reservationId);
+    }
+}
```
검토할게

YehyeokBang commented on src/main/java/roomescape/presentation/api/reservation/request/CreateWaitingRequest.java:
```diff
@@ -0,0 +1,24 @@
+        Long theme,
+        @NotNull(message = "themeId는 필수입니다.")
+        Long time
+) {
```
수정할게

YehyeokBang committed:
fix: DTO 검증 메시지 오타 수정

YehyeokBang committed:
fix: 에러 alert 메시지 오타 수정

yebink commented on src/main/java/roomescape/domain/reservation/Waiting.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    private boolean isOwner(Member member) {
+        return this.member.equals(member);
```
보통 엔티티 비교는 equals 보다는 id 비교를 많이 사용합니다.
id 는 보통 unique 하기 때문에 보다 빠르게 확인할 수 있고,
equals 로 비교하게 되면 새로운 연관관계가 추가되었을 때 해당 엔티티까지 equals를 구현해줘야 하는 번거로움이 있습니다 😄

YehyeokBang replied:
리뷰 감사합니다! 👍 

동의합니다. `Member` 객체 전체를 비교하는 것보다, id를 통해 비교하는 편이 더 가볍고 안정적이며 유지보수 측면에서도 유리할 것 같네요.

또한, `equals()`에 새로운 필드가 포함되거나 연관관계가 추가되는 경우 비교 조건이 변경될 수 있는 점도 고려하면, id 기반 비교가 더 명확하다고 느껴지네요.

반영하겠습니다! 😄 

[refactor: 대기 주인 확인 시 equals 대신 ID 비교로 변경](https://github.com/woowacourse/spring-roomescape-waiting/pull/261/commits/842878e7568207bdadf206b22238267e6e1cafc6)

yebink commented on src/main/java/roomescape/application/auth/LoginService.java:
```diff
@@ -2,39 +2,43 @@
 @Service
 @Transactional(readOnly = true)
-public class AuthService {
+public class LoginService {
```
변경된 이름이 훨씬 명확하네요 👍

yebink commented on src/main/java/roomescape/application/reservation/command/WaitingPromotionService.java:
```diff
@@ -0,0 +1,98 @@
+        Waiting waiting = getWaiting(waitingId);
+        validateControlPermission(member, waiting);
+        validateReservationIsAvailable(waiting);
+        validateWaitingIsFirst(waiting);
```
수동 승인으로 구현되어 있으니 `첫 번째 대기만 승인 가능`할 필요는 없지 않을까요?!

YehyeokBang replied:
<img width="1025" alt="image" src="https://github.com/user-attachments/assets/7c65324b-4302-477f-85bb-88febcc0fdfa" />

현재 주어진 화면에서는 `모든 대기`를 단순하게 보여주고 있으며,
어떤 대기라도 모두 승인 및 거절이 가능한 구조였습니다.
(1~3행은 모두 같은 예약에 대한 대기이며 모두 버튼이 노출되는 구조)

그러다 보니 특정 예약의 첫 번째가 아닌 대기라도 승인을 누를 수 있었습니다.
그래서 검증을 추가하게 되었습니다.

> 조회 쿼리 오류로 첫 번째가 아닌 대기를 응답하거나, 프론트 오류로 첫 번째 대기가 아닌 대기에 버튼이 활성화된다면, `id로 웨이팅을 승인하는 API`를 호출할 때, 승인이 될 수 있겠다는 고민을 했습니다.

검증이 없으면 문제가 발생할 것 같다고 생각해서 더 엄격하게 확인하는 로직을 구현했었습니다.
(승인 요청하신 대기는 첫 번째가 아니라서 승인할 수 없다고 더 자세한 예외 메시지를 응답하는 것도 도움이 될 것 같네요.)

화면을 수정하여, 같은 대기끼리 줄을 세우는 UI로 변경하거나
자동 승인을 구현하게 된다면, `선택한 대기가 첫 번째인지 확인하는 로직`은 없어도 될 것 같네요.

YehyeokBang replied:
[feat: 예약 취소 시 대기 자동 승인 기능 구현](https://github.com/woowacourse/spring-roomescape-waiting/pull/261/commits/a44ecfe9180a8f896ca026708b673564df8d687f)

자동 승인을 구현하게 되면서, `첫 번째 대기만 승인 가능 검증`을 제거했습니다! 👍

yebink replied:
아하! 그런 이유였군요 이해했습니다 ㅎㅎ 👍

yebink commented on src/main/java/roomescape/application/reservation/command/CreateReservationTimeService.java:
```diff
@@ -0,0 +1,32 @@
+
+    private void validateAlreadyExistsReservationTime(CreateReservationTimeCommand createReservationTimeCommand) {
+        if (reservationTimeRepository.existsByStartAt(createReservationTimeCommand.startAt())) {
+            throw new ReservationTimeException("이미 존재하는 얘약시간입니다.");
```
```suggestion
            throw new ReservationTimeException("이미 존재하는 예약시간입니다.");
```

YehyeokBang replied:
[fix: 이미 존재하는 예약 검증 메시지 오타 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/261/commits/79acaccb82098a644f7db249afb4064a3c7f767f)

감사합니다! 👍

yebink commented on src/main/java/roomescape/domain/reservation/repository/WaitingRepository.java:
```diff
@@ -0,0 +1,55 @@
+                 WHERE w2.theme = w.theme
+                   AND w2.date = w.date
+                   AND w2.time = w.time
+                   AND w2.id < w.id))
```
`Waiting`의 id로 웨이팅 순서를 비교하고 있네요.
만약 웨이팅 변경이 가능해져서, 예약 시간 또는 테마를 바꾸게 된다면 어떻게 될까요?

YehyeokBang replied:
## 고민하며 느낀점

**"id는 단순히 각 레코드를 식별하는 것에만 책임져야 한다고 생각해요."**

그러나 현재 구조는 id 순으로 정렬되기 때문에
1. 생성 시간(createAt)보다 정렬 기준이 부정확할 수 있고,
2. PK 전략이 Identity에서 Sequence로 바뀌면 순서가 보장되지 않습니다.

결국 식별자가 식별 이외에 순서의 책임까지 수행하기 때문에 문제가 발생할 수 있는 것 같습니다.

시간 순으로 정렬되는 것이 정확한 의도이기 때문에 비교 기준을 `createdAt`로 수정했습니다.

## 기타 생각

실제 줄을 서는 것처럼 대기를 수정하기 보다는
대기열에서 빠져 나온(삭제) 후 다시 돌아와서 줄을 서는(추가) 것이 자연스럽다고 생각하여 `대기가 수정되는 일은 없다고 가정`했습니다.

만약 대기가 수정되는 요구사항이 추가된다면 기준을 다시 고민해봐야겠네요.
- 예: 최종 수정 일자 등

감사합니다! 👍 

[refactor: 대기 순번을 id 대신 생성일자로 확인하도록 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/261/commits/f279a43d14ce8682e1b27ff082eebd99c5f90a3f)

yebink commented on src/main/java/roomescape/infrastructure/error/GlobalExceptionHandler.java:
```diff
@@ -46,24 +111,9 @@ public ResponseEntity<ApiFailResponse> handleHttpMessageNotReadableException(Htt
-        return ResponseEntity.status(HttpStatus.FORBIDDEN).body(new ApiFailResponse("접근권한이 없습니다."));
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<ApiFailResponse> handleException(Exception e) {
+        log.error("예상치 못한 에러 발생", e);
```
log 활용 👍

yebink reviewed (changes requested):
안녕하세요 벨로~

3-4단계도 잘 구현해주셨네요 👍 
몇 가지 피드백 남겨두었으니 확인해주세요 😄

YehyeokBang committed:
fix: 이미 존재하는 예약 검증 메시지 오타 수정

YehyeokBang committed:
refactor: 대기 제어 확인 메서드명 수정

- 권한 주체와 객체의 위치가 어색하지 않도록 메서드명을 수정합니다.

YehyeokBang committed:
refactor: 대기 주인 확인 시 equals 대신 ID 비교로 변경

- 기존 엔티티를 equals로 비교하던 방식에서 id를 비교하도록 변경합니다.

YehyeokBang committed:
fix: 테스트 검증 메시지 오타 수정

- 논리 오류도 수정합니다.

YehyeokBang committed:
refactor: 대기 순번을 id 대신 생성일자로 확인하도록 수정

YehyeokBang committed:
feat: 예약 취소 시 대기 자동 승인 기능 구현

- 예약 취소 후 이벤트 발행 방식으로 대기 자동 승인 처리합니다.
- EventListener 기반으로 결합도 낮추고 도메인 독립성을 유지합니다.

YehyeokBang committed:
refactor: 코드 정렬

- import 문 순서를 조정합니다.
- 빼먹은 접근 지정자를 추가합니다.

YehyeokBang reviewed (commented):
안녕하세요 엘리!

이번 리뷰도 생각 정리에 많은 도움이 되었습니다.
감사합니다. 👍 

덕분에 자동 승인도 도전해보게 되었네요.
이번 리뷰도 잘 부탁드립니다. 😄

yebink reviewed (approved):
안녕하세요 벨로~

피드백 잘 반영해주셨네요 😄 
더 이상 드릴 피드백이 없네요. 이번 미션 고생 많으셨습니다! 💯
